runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PTransformMatchers.java

```diff
@@ -106,6 +106,64 @@ public String toString() {
     }
   }
 
+  /**
+   * A {@link PTransformMatcher} that matches a {@link ParDo.SingleOutput} containing a {@link DoFn}
+   * that requires stable input, as signified by {@link ProcessElementMethod#requiresStableInput()}.
+   */
+  public static PTransformMatcher requiresStableInputParDoSingle() {
+    return new PTransformMatcher() {
+      @Override
+      public boolean matches(AppliedPTransform<?, ?, ?> application) {
+        PTransform<?, ?> transform = application.getTransform();
+        if (transform instanceof ParDo.SingleOutput) {
+          DoFn<?, ?> fn = ((ParDo.SingleOutput<?, ?>) transform).getFn();
+          DoFnSignature signature = DoFnSignatures.signatureForDoFn(fn);
+          return signature.processElement().requiresStableInput();
```

Comment:
Is it possible to annotate `@OnTimer` method with `requiresStableInput` without annotating `ProcessElement`?

Reply:
Is it possible to save this information for the `DoFn`, instead of the `ProcessElement`? So that it will just be `signature.requiresStableInput`? I may be wrong

Reply:
1) Yes. Actually the prototype for the annotation on @OnTimer method is already there, but I will leave the support of that for future work. See https://github.com/apache/beam/blob/279a05604b83a54e8e5a79e13d8761f94841f326/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignature.java#L676.
2) I think we previously decided on that the annotation should be on @ProcessElement and @OnTimer method. Therefore, the prototype is written to reflect decision. See the previous link for more details. Thanks for the comments.